## Упражнения по библиотеке Numpy

In [3]:
import itertools
import random
from statistics import median

import numpy as np


**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [5]:
array = np.random.randint(0, 10, size=30)
array = np.where((array > 3) & (array < 8), -array, array)
array

array([-4, -6,  8,  0,  3,  3,  9,  2,  2,  8,  8, -7,  0,  0,  9, -7,  3,
        2, -5, -7, -4,  1,  8,  2,  3, -6,  0,  1,  0,  0], dtype=int32)

**2.** Заменить максимальный элемент случайного массива на 0

In [6]:
array = np.random.randint(0, 100, size=10)
array = np.where(np.max(array) == array, 0, array)
array

array([25, 72,  0, 73, 55, 54, 66, 71, 67, 75], dtype=int32)

**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [7]:
array = np.array([(1, 2, 3), (4, 5, 6), (31, 333, 1)])
one = np.copy(array)
result = one * array
result

array([[     1,      4,      9],
       [    16,     25,     36],
       [   961, 110889,      1]])

In [8]:
array = np.array([(1, 2, 3), (4, 5, 6), (31, 333, 1)])
it = list(itertools.product(*array))
array = np.array(it)
array

array([[  1,   4,  31],
       [  1,   4, 333],
       [  1,   4,   1],
       [  1,   5,  31],
       [  1,   5, 333],
       [  1,   5,   1],
       [  1,   6,  31],
       [  1,   6, 333],
       [  1,   6,   1],
       [  2,   4,  31],
       [  2,   4, 333],
       [  2,   4,   1],
       [  2,   5,  31],
       [  2,   5, 333],
       [  2,   5,   1],
       [  2,   6,  31],
       [  2,   6, 333],
       [  2,   6,   1],
       [  3,   4,  31],
       [  3,   4, 333],
       [  3,   4,   1],
       [  3,   5,  31],
       [  3,   5, 333],
       [  3,   5,   1],
       [  3,   6,  31],
       [  3,   6, 333],
       [  3,   6,   1]])

**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [9]:

A = np.random.randint(0, 10, size=(8, 3))
B = np.random.randint(0, 10, size=(2, 2))

sep = (A[..., np.newaxis, np.newaxis] == B)
row = (sep.sum(axis=(1, 2, 3)) >= B.shape[1]).nonzero()[0]
row

array([2, 3, 4, 5, 7])

**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [10]:
arr1 = np.array([
    [1, 1, 1],
    [2, 2, 2],
    [7, 8, 9],
    [0, 1, 2],
    [3, 4, 5],
    [6, 7, 8],
    [9, 0, 1],
    [2, 3, 4],
    [5, 6, 7],
    [8, 9, 0]
])
m = np.logical_and.reduce(arr1[:, 1:] == arr1[:, :-1], axis=1) & np.logical_and.reduce(arr1[:, 1:] == arr1[:, [0, 2]],
                                                                                       axis=1) & np.logical_and.reduce(
    arr1[:, :-1] == arr1[:, [0, 2]], axis=1)
arr1[~m]


array([[7, 8, 9],
       [0, 1, 2],
       [3, 4, 5],
       [6, 7, 8],
       [9, 0, 1],
       [2, 3, 4],
       [5, 6, 7],
       [8, 9, 0]])

**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [11]:
arr1 = np.array([[1, 1, 1], [2, 2, 2], [1, 1, 1]])
arr_uniq = np.unique(arr1)

arr1, arr_uniq


(array([[1, 1, 1],
        [2, 2, 2],
        [1, 1, 1]]),
 array([1, 2]))

______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [12]:

X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]])
list_x = X.tolist()

mult = 1
for i in range(len(list_x[0])):
    if X[i][i] != 0:
        mult *= X[i][i]

mult


np.int64(3)

In [13]:

X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]])
result = np.prod(np.diag(X)[np.diag(X) != 0])
result

np.int64(3)

* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [14]:

x = [1, 2, 2, 4]
y = [4, 2, 1, 2]


def count_el(arr):
    count_d = {}
    for elenent in arr:
        if elenent in count_d:
            count_d[elenent] += 1
        else:
            count_d[elenent] = 1
    return count_d


def equals(x, y):
    count_x = count_el(x)
    count_y = count_el(y)

    return count_x == count_y


result = equals(x, y)
result





True

In [20]:

x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 2])
result = True
unique_x, counts_x = np.unique(x, return_counts=True)
unique_y, counts_y = np.unique(y, return_counts=True)

if np.array_equal(counts_x, counts_y) and np.array_equal(unique_x, unique_y):
    result = True
else:
    result = False
result




15.3 µs ± 165 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [15]:

x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])
x = x.tolist()
p = max(x[i] for i in range(len(x)) if i != 0 if x[i - 1] == 0)
p

1.25 µs ± 13.9 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


In [77]:

x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])
res = np.max(x[1:][(x == 0)[:-1]])
res

np.int64(5)

* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [70]:
def run_length_encoding(x):
    values = []
    count = []
    current_value = x[0]
    current_count = 1

    for i in range(1, len(x)):
        if x[i] == current_value:
            current_count += 1
        else:
            values.append(current_value)
            count.append(current_count)
            current_value = x[i]
            current_count = 1

    
    values.append(current_value)
    count.append(current_count)
    
    return (values, count)

x = np.array([2, 2, 2, 3, 3, 3, 5])
x = x.tolist()
run_length_encoding(x)



([2, 3, 5], [3, 3, 1])

In [14]:
x = np.array([2, 2, 2, 3, 3, 3, 5])
def run_length_encoding_numpy(arr1):
    d = np.where(arr1[:1] != arr1[1:])[0]
    print(d)

run_length_encoding_numpy(x)

[2 3 4 5]


* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [1]:
def diff(a, b):
    result = []
    for i in range(a.__len__()):
        result.append(a[i] - b[i])
    return result

In [4]:
%%time
from math import sqrt

a = np.random.randint(10, size=(250, 4))
b = np.random.randint(10, size=(240, 4))

result_list = []
for i in a:
    row = []
    for j in b:
        row.append(np.sqrt(sum(pow(a - b, 2) for a, b in zip(i, j))))
    result_list.append(row)

result_list

CPU times: total: 62.5 ms
Wall time: 156 ms


[[np.float64(4.0),
  np.float64(9.273618495495704),
  np.float64(3.872983346207417),
  np.float64(4.58257569495584),
  np.float64(5.5677643628300215),
  np.float64(3.4641016151377544),
  np.float64(8.306623862918075),
  np.float64(3.0),
  np.float64(5.0990195135927845),
  np.float64(4.795831523312719),
  np.float64(4.47213595499958),
  np.float64(6.557438524302),
  np.float64(6.082762530298219),
  np.float64(8.717797887081348),
  np.float64(9.38083151964686),
  np.float64(6.48074069840786),
  np.float64(6.782329983125268),
  np.float64(6.6332495807108),
  np.float64(4.795831523312719),
  np.float64(6.164414002968976),
  np.float64(5.0),
  np.float64(5.477225575051661),
  np.float64(3.7416573867739413),
  np.float64(6.782329983125268),
  np.float64(2.449489742783178),
  np.float64(4.795831523312719),
  np.float64(5.0990195135927845),
  np.float64(5.385164807134504),
  np.float64(7.0710678118654755),
  np.float64(5.5677643628300215),
  np.float64(8.774964387392123),
  np.float64(8.124038

In [5]:
np.sqrt(np.array(sum([pow(np.array(diff(x, y)), 2) for x, y in zip(a, b)])))

array([61.99193496, 64.49806199, 63.1110133 , 62.92853089])

In [6]:
%%time

a = np.random.randint(10, size=(2250, 4))
b = np.random.randint(10, size=(2240, 4))

P = np.add.outer(np.sum(a ** 2, axis=1), np.sum(b ** 2, axis=1))
N = np.dot(a, b.T)
print(np.sqrt(P - 2*N))

[[ 8.83176087 11.61895004  4.58257569 ...  6.70820393  6.70820393
   7.68114575]
 [ 9.69535971  9.74679434  7.54983444 ...  8.54400375  5.38516481
   8.18535277]
 [ 6.8556546  11.40175425  6.164414   ...  7.61577311 10.
   7.61577311]
 ...
 [ 9.32737905  9.2736185   6.         ...  8.71779789  3.74165739
   8.1240384 ]
 [ 8.36660027  8.42614977  5.91607978 ...  8.88819442  4.79583152
   7.93725393]
 [ 7.34846923 13.07669683  5.19615242 ...  2.64575131  9.32737905
   4.35889894]]
CPU times: total: 0 ns
Wall time: 55.7 ms


In [7]:
%%time
import scipy

a = np.random.randint(10, size=(22500, 4))
b = np.random.randint(10, size=(22400, 4))

scipy.spatial.distance.cdist(a, b)

CPU times: total: 500 ms
Wall time: 1.85 s


array([[10.14889157, 12.72792206,  7.07106781, ...,  9.48683298,
        10.29563014, 11.74734012],
       [ 3.        ,  6.164414  ,  5.09901951, ...,  6.        ,
         4.69041576,  5.29150262],
       [ 5.38516481,  8.48528137,  3.16227766, ...,  7.61577311,
         8.36660027,  4.69041576],
       ...,
       [ 6.55743852, 12.24744871,  5.09901951, ...,  6.92820323,
         5.47722558,  8.24621125],
       [ 8.66025404,  7.07106781,  9.48683298, ...,  8.1240384 ,
         9.89949494,  8.83176087],
       [ 4.79583152,  5.47722558,  6.        , ...,  8.83176087,
         7.21110255,  7.34846923]])

_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [1]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [2]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [3]:
average_calories = np.mean(calorie_stats, axis=0)
diff_calories = average_calories - 60
print(f"Среднее значение по конкурентам: {average_calories}, Разница среднего и нашей компании: {diff_calories}")

Среднее значение по конкурентам: 106.88311688311688, Разница среднего и нашей компании: 46.883116883116884


3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [4]:
calorie_stats_sorted = np.sort(calorie_stats)
print(f"Отсортированные данные: {calorie_stats_sorted}")

Отсортированные данные: [ 50.  50.  50.  70.  70.  80.  90.  90.  90.  90.  90.  90.  90. 100.
 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100.
 100. 100. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 120. 120. 120. 120. 120. 120. 120. 120. 120. 120. 130.
 130. 140. 140. 140. 150. 150. 160.]


4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [5]:
median_calories = np.median(calorie_stats)
print(f"Медиана: {median_calories}")


Медиана: 110.0


5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [6]:
for pr in range(1, 101):
    value = np.percentile(calorie_stats, pr)
    print(f"Процентиль номер:{pr} и значением : {value}")
    if value > 60:
        nth_percentile = pr
        print(f"Наименьший процентиль, превышающий 60 калорий, это {pr}-й процентиль с значением {value} калорий.")
        break
    


Процентиль номер:1 и значением : 50.0
Процентиль номер:2 и значением : 50.0
Процентиль номер:3 и значением : 55.599999999999994
Процентиль номер:4 и значением : 70.0
Наименьший процентиль, превышающий 60 калорий, это 4-й процентиль с значением 70.0 калорий.


6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [7]:
more_calories = len(calorie_stats_sorted[calorie_stats_sorted > 60]) / len(calorie_stats_sorted)
print(f"Процент хлопьев в который содержится более 60ккал составляет- {more_calories*100:.2f}%")

Процент хлопьев в который содержится более 60ккал составляет- 96.10%


7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [8]:
#это по формуле
squared_diffs = [(value-average_calories)  ** 2 for value in calorie_stats_sorted]
avg_sqr = sum(squared_diffs) / len(calorie_stats_sorted)
calorie_std = avg_sqr ** 0.5
print(calorie_std)

    

19.357185333908266


In [9]:
calorie_std = np.std(calorie_stats)
print(calorie_std)

19.35718533390827


Величина отклонения 19.36 это говорит о том что данные достаточно сильно различаются, что означает что некоторые хлопья(в нашем случае) могут быть сильно калорийными, а другие наоборот. Данные разбросаны, и медиана не дает точного представления о данных.

8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

In [10]:
На основе данных можно сделать вывод, что наши хлопья содержат меньше калорий, чем у 96% конкурентов. Это отличная возможность для отдела маркетинга привлечь внимание любителей хлопьев, а также тех, кто заботится о своем питании и стремится поддерживать фигуру

SyntaxError: invalid syntax (3830361763.py, line 1)